### ISSUE TITLE
Strange conflict between FillIItUp and this mod

### ISSUE THREAD

**Issue by linuxgurugamer** [SteveBenz/ProgressiveColonizationSystem#2]
I'm looking into a strange issue when both FillItUp and ProgressiveColonication are installed into a 1.9.1 game. I do not believe it happens in 1.8.1, but haven't tested that yet.

The problem happens after going from the editor (VAB) to launching a vessel.  The stock toolbar has the stock icons duplicated as seen in the attached picture
![screenshot1](https://user-images.githubusercontent.com/1630166/81632375-0daecd00-93d8-11ea-91a9-044d8da38b6d.png)

Would you have any idea what's going on?

I've been able to reduce the problem to ONLY these two mods, no other mods installed  other than what is needed for FillItUp.
![explorer_2020-05-11_22-39-01](https://user-images.githubusercontent.com/1630166/81632452-5070a500-93d8-11ea-9738-569ce5d01509.png)

I'll be looking into this, would appreciate any insights you may have

**Comment by pandrien** [comment]
I've seen this bug before. I installed some kind of tool bar mod and that caused it. I'm sorry I forget which one. I uninstalled the toolbar mod and it went away. I think we can't make the assumption that the bug is limited to FillItUp or ProgressiveColonizationSystem.

Edit: I went back into my logs and I strongly suspect that it's "ToolbarControl" at fault.

**Comment by pandrien** [comment]
Ok I just did a test. I installed Progressive colonization and Fill It Up. I also let CKAN do it's thing and install the dependencies for Progressive colonization. 

![image](https://user-images.githubusercontent.com/392937/81662308-3ba10b00-93f2-11ea-85a3-23783d73ada0.png)

No repo on the bug. I strongly suspect that the culprit is "ToolbarControl" (not to be confused with Toolbar Controller which is fine)

I would suggest you remove ToolbarControl and confirm if that fixes it.

**Comment by pandrien** [comment]
Just to clarify I'm not sure there is actually a separate mod called toolbarcontrol but i know at one point I had two separate entries, one called "Toolbar Controller" and another called "toolbarControl" in ckan AND I also had the double toolbar like you are seeing.

Regardless I think this is not an issue with Progressive Colonization System.

**Comment by linuxgurugamer** [comment]
The ToolbarController is a mod that I wrote to manage access to both the stock toolbar and the Blizzy toolbar, with a single API.
I'll try installing the other dependencies of PC and see what happens.

While I doubt it's the toolbar, I will include it in my investigations.

To be blunt, I don't know what's causing this.  I only contacted you because PC was one of the two mods which seemed to trigger the bug, and to solicit your input.

I have downloaded PC and can locally build it, so I'll be able to track this down, whereever it may be occuring.

Re. removing the ToolbarController, if I do, then FillItUp won't work, it's a dependency.  However, for testing purposes, I'll see if I can remove it from a branch and just put in a simple stock toolbar button,  it's a good suggestion.

Just so you know, all of my mods which require a toolbar use the ToolbarController, and I regularly play a game with over 200 mods installed, at least half of which use the ToolbarController.

I'll keep you updated

**Comment by linuxgurugamer** [comment]
> Just to clarify I'm not sure there is actually a separate mod called toolbarcontrol but i know at one point I had two separate entries, one called "Toolbar Controller" and another called "toolbarControl" in ckan AND I also had the double toolbar like you are seeing.
> 
> Regardless I think this is not an issue with Progressive Colonization System.

Just reread your comment, wanted to clarify:

1.  "toolbarControl" is the name of the variable I use in all my mods to hold an instance of the ToolbarController (case is accurate)
2.  I just added back all the missing mods which CKAN normally installs, the bug happened again.
3.  I realized that it's quite possible the bug is related to something in my save file, so I created a new save to test.  This was a new sandbox save, it happened again.
4.  Did you test with a career save or a sandbox save?
5.  I then deleted the settings.cfg (renamed, actually) and started a new sandbox save, happened again.

Lastly, I'm curious as to how you started the game.  Specifically, what are the startup options you passed in.  I'm not sure that's relevant, but with two datapoints (yours and my systems), I'm looking for the differences.

Also, could you send me the log file of the test you did?  I want to verify that we are using the same versions.  Again, looking for datapoints.

Thanks

**Comment by pandrien** [comment]
For testing I opened a career save that I had. [settings.cfg.txt](https://github.com/SteveBenz/ProgressiveColonizationSystem/files/4616702/settings.cfg.txt). I'd rather get a clean install before sending a log file. 

I just installing Toolbar Controller and was able to reproduce the bug.
![image](https://user-images.githubusercontent.com/392937/81714320-ce11d080-942b-11ea-8827-4705336329b9.png)

KSP version is 1.9.1.2788. Mod versions are in the screenshot.

**Comment by pandrien** [comment]
Here is a clean log [CleanLogs.zip](https://github.com/SteveBenz/ProgressiveColonizationSystem/files/4616871/CleanLogs.zip)

Steps:
1. Deleted KSP install and did file integrity check in steam to reinstall fresh.
2. Created copy for testing and loaded up CKAN
3. Installed only Fill It Up and Progressive Colonization (Toolbar and Toolbar Controller are not dependencies). (enabled 1.9 and 1.8 as compatible versions)
4. Load KSP and opened new sandbox save.

This is the version with **no repo**.

**Comment by linuxgurugamer** [comment]
With thi scleanLogs, did the bug show up or not?
Re. the one where you reproduced it above, could you zip up the entire save and send that as well.  It's a good datapoint.
And, while I don't think it's significant, it's interesting that you are using Steam, and I am not.  I also have a Steam install, so I can test both ways.
Thanks

**Comment by linuxgurugamer** [comment]
Ummm, if Toobarcontroller isn't installed, FillItUp wont' work.  But that's ok, if I can replicate the problem which you showed above, that woudl remove FillitUp from the mix, I'm trying that now

**Comment by pandrien** [comment]
The bug did not show up. That's what I mean by no repo. Here is the save file. I really don't think that save matters though.
[default.zip](https://github.com/SteveBenz/ProgressiveColonizationSystem/files/4617002/default.zip)

**Comment by linuxgurugamer** [comment]
> For testing I opened a career save that I had. [settings.cfg.txt](https://github.com/SteveBenz/ProgressiveColonizationSystem/files/4616702/settings.cfg.txt). I'd rather get a clean install before sending a log file.
> 
> I just installing Toolbar Controller and was able to reproduce the bug.

Yes, but you also had FillitUp installed.  Fill-It-Up has the ToolbarController as a dependency, if it isn't there, then Fill-It-Up won't run.  By installing the TC, you also  enabled Fill-It-Up, so this doesn't change anything.  But having the log file will be very helpful.  I'm working right now, hope to be able to do a deep debugging session later this evening.

**Comment by linuxgurugamer** [comment]
Just an FYI, not related to this.  I noticed from the log that you still have one or more MiniAVC.dll files, you should remove them all.  It is seriously broken in 1.8 and 1.9 and is not being used anymore.  You can install ZeroMiniAVC which will do that automatically for you

**Comment by pandrien** [comment]
Ok so I tested a bunch of different combinations. I agree it's very strange. 
PCS + FIU = no bug
PCS + TC = no bug
PCS + FIU + TC = BUG

Regarding MiniAVC, I see one that gets installed by Fill It Up but that's it.
![image](https://user-images.githubusercontent.com/392937/81740580-09bf9100-9452-11ea-9ecc-53b62e4ccfb9.png)

Also you say that Fill It Up has ToolbarController as a dependency but it's not recorded that way in CKAN for me:
![image](https://user-images.githubusercontent.com/392937/81740774-5acf8500-9452-11ea-8e51-f6a710dcd4bd.png)

P.S. I am not a mod author just an enthusiastic user if that wasn't clear.

**Comment by linuxgurugamer** [comment]
Oh nuts, I'll have to get CKAN updated.  Thanks

**Comment by SteveBenz** [comment]
This is new to me, that's for  sure!  Here's the code that deals with the toolbar:

https://github.com/SteveBenz/ProgressiveColonizationSystem/blob/628af67c782c5e338690d2bd2df2e929661dd5ec/src/ProgressiveColonizationSystem/PksToolbarDialog.cs#L62

This was some of my earliest work, and I struggled quite a bit with getting this button to work correctly; I wouldn't be surprised if I made a mistake here.

**Comment by linuxgurugamer** [comment]
I've spent the day on this, and don't have a great answer, other than I see that you are using multiple ScenarioModules, one of which (at least) is inherited by some child classes.

Specifically, the PksToolbarDialog.  When I change it from a ScenarioModule to a MonoBehaviour, the problem goes away, which leads me to think there is an issue in the ScenarioModule.  I then looked into the classes which used PksToolbarDialog, and noticed that they weren't using the features of the ScenarioModule, so, at least to me, it looks like it isn't needed.

Of course, you know your code better than I do.

One other thing, as part of my investigation, I switched it over from using the stock  AddModApplication() to using the ToolbarController.  

So, can you look to see if the ScenarioModule can be replaced with MonoBehaviour for the PksToolbarDialog()?

And, let me know if you'd like the code changes to use the ToolbarController

**Comment by linuxgurugamer** [comment]
The problem of multiple buttons when both Fill-It-Up and Progressive-Colonization are present (and most likely Werner's Checker as well) arises because both mods are replacing the listener of the EditorLogic.fetch.launchBtn.onClick.  Then, when the button is clicked, BOTH replacements are called, and then both of them are invoking the EditorLogic.fetch.launchVessel.

This isn't a bug in either mod, it's a bug in the implementation since there is no way for a mod to get a delegate of a button which is an unknown method.

While it's possible to remove all listeners, the only one this would work for would be the last mod initialized, since it would remove all others.

Ideally, what should happen is one of the following (for the purposes of this discussion, I'll refer to both mods via abbreviations, and also assume there are only two mods, but this applies to any number of mods):

1.  The new UnityAction which each mod adds would be called simultaneously, and then only when both have finished, the original delegate (EditorLogic.fetch.launchVessel) would be called.

2.  The first mod initialized would be called first, and when it was done the second mod would be called, and finally the original delegate would be called.

3.  A mod will specify a priority when initializing .


This can't be solved by one mod, a solution needs to be implemented and used by all mods which replaces the delegate(s) of this button.

I am working on a solution for this because:
a.  I found the problem
b.  I have at least two mods affected by this problem

I'm writing a tiny mod which will be used by any other mod to control access to this button and it's delegates.

class ButtonManager
{
	void InitializeListener(UnityEngine.UI.Button button,  UnityAction delegate);
	int AddListener(UnityEngine.UI.Button button,  UnityAction delegate, int priority = 5);
	void RemoveListener(UnityEngine.UI.Button button,  UnityAction delegate);
	void RemoveListener(int listenerId);

	int GetListenerCount(UnityEngine.UI.Button button);
	void RemoveAllListeners(UnityEngine.UI.Button button);

	void InvokeNextDelegate(int currentDelegateId);

	void SetModeSequential();
	void SetModeSimultaneous();
}

The priority will be used to determine the order of listeners to be called.  Higher priorities will be run first.  For the Sequential mode, mods with the same priority will be run in the order initialized.

InitializeListener	Initializes the mod to work with the specified button.  The delegate should be the same as what the default it, it will have the lowest priority
AddListener	Adds a delegate to a button with a specified priority.  Returns a listenerId, used below
RemoveListener	Removes a delegate from a button
GetListenerCount	Gets a count of all the listeners
RemoveAllListeners	Removes all except the default listener
InvokeNextDelegate	Tells the mod to call the next delegate on the list

SetModeSequential		These two can set or change the mode, sequential will run one
SetModeSimultaneous	delegate at a time, simultaneous will run all of the same priority at the same time, and then the next lower, etc


I hope to have this finished in the next few days, and will first implement it in my mods.  I'd be happy to do the same for ProgressiveColonization if you'd like.

**Comment by linuxgurugamer** [comment]
I have it finished, and doing testing now.  Getting back to your mod, I can (if you like) do it for you with the following options:

1. By itself, no other changes
2. This fix, plus the changes for the ToolbarController

Please let me know ASAP which, if any, you'd like.  Which ever I do, will be done by commenting out your code so that you can specifically see the changes that I did

**Comment by SteveBenz** [comment]
Sorry I haven't been responsive on this -- been pretty wrapped up at work.

I'm always fond of granular changes, so if there were separate PR's, that'd be nicer.  Don't comment out my code - just delete what you don't think should remain.  Let git record what the old stuff was.

Reading over the proposed ButtonManager class...  I never like generic stuff like "priority" because what makes for a good priority?  Unless I know what the other mods have chosen, I'm kinda in the dark...  Which means I **am** in the dark.  But there's no other way to go from  your perspective either since you do know that some mods need to go before others.  I think, however, that maybe there are common things that drive the choice -- like does it rewrite the ship or not, does it abort the launch or not.  Rewriters should come before abort choices.  (So maybe the documentation could suggest certain values for these?)


As to toolbars, as a first-time Mod-Author, the developer experience of adding a toolbar (in stock) was awful.  What there should be is a single API with these arguments:

1. The scenes within which the button will be active
2. The highlighted/unhighlighted texture
3. A per-frame callback for when it's active
4. A comment that explains why it's a per-frame callback and not an on-click like most folks are used to.

It'd be nice if there was a mod-library that helped with that.

**Comment by linuxgurugamer** [comment]
Ok, to address each point:

The priority will default to 5 (1-10 range), higher is done first, and yes, there will be documentation.  There will also be a UI for people to change it, so my overall feeling about it is to just leave it unless there is a big reason, and let the players decide. 

Regarding the toolbar, I agree with you, which is why I wrote the ToolbarController.  Documented here:
https://forum.kerbalspaceprogram.com/index.php?/topic/169509-19x-toolbar-controller-for-modders/

It is a single call instead of the mess that stock and the Blizzy Toolbar is, and additionally supports both the stock and the Blizzy Toolbar.
I'm not sure what you mean by a "per-frame" callback, it's a button which does a "on-click" callback to a single method which gets passed in, it can also support the mouse onHover, onHoverOut, and also onEnable and onDisable, essentially the same that the stock toolbar calls.  

Ok.  I can get you the toolbar code in a day or so, and the ButtonManager code I can do when I finish and release it.

**Comment by linuxgurugamer** [comment]
I sent you a PR earlier with the changes for the ToolbarController

**Comment by linuxgurugamer** [comment]
Hi,
I released the ButtonManager today along with the ClickThroughBlocker update and a bunch of others.  I've updated  my mod, FillItUp to use the buttonmanager as well.
I'll be updating the other mod I have which also takes control of the launch button, WernherChecker on Monday.  It has some dependencies which I want to wait on, since I'm running out of time.

I submitted a PR to you for the ToolbarController, I see you haven't merged it yet.

I'll pull a fresh copy of this mod and update it for the ButtonManager, and then submit that PR to you as well.  That will probably be on Monday, unless I get some time earlier.
